nixos/modules/services/audio/mpd.nix

```diff
@@ -158,18 +158,18 @@ in {
       };
     };
 
+    systemd.tmpfiles.rules = [
+      "d '${cfg.dataDir}' - ${cfg.user} ${cfg.group} - -"
+      "d '${cfg.playlistDirectory}' - ${cfg.user} ${cfg.group} - -"
+    ];
```

Comment:
I don't think this is the right way either. I'm sure it will be working but it has multiple issues as well:
* the name: surely the dataDir shouldn't be treated like a tmpdir
* the location: this service now includes all of the rules of all of the services. It's going to be hard to determine what rule is used for what service. Looking at the mdp unit won't give much clue to the ties with this service.

Reply:
adding @flokli and @fpletz to the conversation

Reply:
@zimbatm
what if `systemd.services` got a new suboption - `ensureDirs`, which is implemented as `systemd.tmpfiles` (for NixOS) or shell script (for service runner) under the hood? Would that address your concerns?

Reply:
@zimbatm unfortunately, it seems my comment about tmpfiles isn't really accessible - see https://github.com/NixOS/nixpkgs/pull/56265#discussion_r264064056 for an explanation why the naming is unlucky, but this fine in general (also the comments afterwards regarding `StateDirectory` and `ConfigurationDirectory`)

Reply:
@danbst partially, at least it would be more explicit in the nix code. But the generated unit itself won't have that information.

from the manual:
> Use RuntimeDirectory= to manage one or more runtime directories for the unit and bind their lifetime to the daemon runtime. This is particularly useful for unprivileged daemons that cannot create runtime directories in /run due to lack of privileges, and to make sure the runtime directory is cleaned up automatically after use. For runtime directories that require more complex or different configuration or lifetime guarantees, please consider using tmpfiles.d(5).

I believe that tmpfiles.d is only really meant for runtime directories (which are temporary)

---

`StateDirectory` should be used whenever possible, but it's not always possible because `cfg.stateDir` allows the user to relocate that directory to another place. That's the crux of the problem really.

What happens if a StateDirectory is given an absolute path?

Reply:
> What happens if a StateDirectory is given an absolute path?

According to `man 5 systemd.exec` this is not intended/supported (i.e. even if it does work we should not rely on it):
> The specified directory names must be relative, and may not include "..".

Which is unfortunate as it would be a nice solution together with `StateDirectoryMode=` IMO.

I'm also a bit concerned about having all rules in `/etc/tmpfiles.d/nixos.conf` (which makes the mapping between `tmpfiles` and the unit file difficult in both directions).

Reply:
A great solution (IMO) is to no longer provide users the option to specify state, configuration, runtime, and log directories. Other distros don't provide this option. It seems to have become a bit of a pattern in nixos to provide a stateDir option just because "we can" (which is not justification for doing so).

Stop giving the option and then you can just use StateDirectory, etc... and we won't have to work against systemd tools because we won't be using them in ways they likely weren't intended to use.

Reply:
Had an offline chat with @zimbatm about that - it should still be possible to mount external storage from other places to `StateDirectory`, `RuntimeDirectory` etc if we really want to - but cluttering all units with "custom state dir awareness logic" was a bad idea in first place.

Reply:
> A great solution (IMO) is to no longer provide users the option to specify state, configuration, runtime, and log directories. Other distros don't provide this option. It seems to have become a bit of a pattern in nixos to provide a stateDir option just because "we can" (which is not justification for doing so).

I don't think this configuration is valid.
First of, other distros usually give the user the option to change configuration anyway they want to via mutable `etc`.
Second, and I think this is the most relevant, other distros do not change configuration under the covers but instead prompt for feedback if they detect differences in configuration that would otherwise be overridden. This is not something we should strive to do, naturally.
Since I consider state THE most important, critical and most difficult part to handle correctly on any running system, taking away control over managing state feels like a really bad idea to me.

Reply:
>     * the name: surely the dataDir shouldn't be treated like a tmpdir

I agree with this. The semantics of the systemd.tmpfiles is clearly towards storage of volatile data, eg.

> systemd-tmpfiles uses this configuration to create volatile files and directories during boot and to do periodic cleanup afterwards.

> d  Create a directory. The mode and ownership will be adjusted if specified. Contents of this directory are subject to time based cleanup if the age argument is specified. 

Using tempfile as a manner to store persistent state is at the very least unexpected by the maintainer of a system. It also feels a lot like a hack that is used not because of correctness but because it's conveniently available. Since this change will impact how user data is stored and handled, relying on the implementation instead of the intention of a systemd feature makes me nervous.

Reply:
@kampka regarding the semantics of tmpfiles: https://github.com/systemd/systemd/issues/12447
TLDR: tmpfiles is perfectly fine for managing stateful things.

> First of, other distros usually give the user the option to change configuration anyway they want to via mutable etc.

This PR is about letting systemd handle the initial creation of `StateDirectory`, `LogsDirectory` and `RuntimeDirectory`, and the fs structure inside (if the service doesn't set it up on its own).
We rely on the "native" systemd primitives here, as they provide a much more declarative way in describing the file structures and permissions than PreStart shell scripts.

This doesn't take away options to provide an alternative location for storing state, the fact we provided these options simply makes things a bit more complex than it should be. As `StateDirectory` needs to be inside `/var/lib/…` we use tmpfiles to set up the state directory  too, when its location could (in theory) be outside of `/var/lib`.

A way to clean that up any further might be to (bind-)mount whatever state directory is chosen to the native `StateDirectory` location, but IMHO thats out of the scope of this PR.

Reply:
> @kampka regarding the semantics of tmpfiles: [systemd/systemd#12447](https://github.com/systemd/systemd/issues/12447)
> TLDR: tmpfiles is perfectly fine for managing stateful things.

I was not aware of that issue, but I agree with it.

> This doesn't take away options to provide an alternative location for storing state, the fact we provided these options simply makes things a bit more complex than it should be. As `StateDirectory` needs to be inside `/var/lib/…` we use tmpfiles to set up the state directory too, when its location could (in theory) be outside of `/var/lib`.

The reason why I joined this discussion was because I is used as a point of reference in other PRs to make a case for not having configurable state directories in the first place (https://github.com/NixOS/nixpkgs/pull/60788#issuecomment-491443263) which is also a topic that was raised in this PR, see https://github.com/NixOS/nixpkgs/pull/56265#discussion_r264065847

As I have a strong opinion about giving users control (options) about how to handle space, my main intend was making that heard, not challenge the scope of the PR.